MapLoader NPE on loadAllKeys()

This code
```
	@Override
	public Iterable<String> loadAllKeys() {
		Collection<String> result = new ArrayList<>();
		result.add(null);
		return result;
	}
```
gives
```
Caused by: java.lang.NullPointerException
	at com.hazelcast.internal.partition.impl.InternalPartitionServiceImpl.getPartitionId(InternalPartitionServiceImpl.java:845) ~[hazelcast-3.8.3.jar!/:3.8.3]
	at com.hazelcast.map.impl.MapKeyLoaderUtil$2.apply(MapKeyLoaderUtil.java:113) ~[hazelcast-3.8.3.jar!/:3.8.3]
	at com.hazelcast.map.impl.MapKeyLoaderUtil$2.apply(MapKeyLoaderUtil.java:110) ~[hazelcast-3.8.3.jar!/:3.8.3]
	at com.hazelcast.util.IterableUtil$2.next(IterableUtil.java:60) ~[hazelcast-3.8.3.jar!/:3.8.3]
	at com.hazelcast.map.impl.MapKeyLoaderUtil.nextBatch(MapKeyLoaderUtil.java:86) ~[hazelcast-3.8.3.jar!/:3.8.3]
	at com.hazelcast.map.impl.MapKeyLoaderUtil$1.next(MapKeyLoaderUtil.java:76) ~[hazelcast-3.8.3.jar!/:3.8.3]
	at com.hazelcast.map.impl.MapKeyLoaderUtil$1.next(MapKeyLoaderUtil.java:65) ~[hazelcast-3.8.3.jar!/:3.8.3]
	at com.hazelcast.map.impl.MapKeyLoader.sendKeysInBatches(MapKeyLoader.java:341) ~[hazelcast-3.8.3.jar!/:3.8.3]
	at com.hazelcast.map.impl.MapKeyLoader.access$200(MapKeyLoader.java:72) ~[hazelcast-3.8.3.jar!/:3.8.3]
	at com.hazelcast.map.impl.MapKeyLoader$1.call(MapKeyLoader.java:191) ~[hazelcast-3.8.3.jar!/:3.8.3]
	at com.hazelcast.map.impl.MapKeyLoader$1.call(MapKeyLoader.java:188) ~[hazelcast-3.8.3.jar!/:3.8.3]
	at com.hazelcast.util.executor.CompletableFutureTask.run(CompletableFutureTask.java:67) ~[hazelcast-3.8.3.jar!/:3.8.3]
	at com.hazelcast.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:228) ~[hazelcast-3.8.3.jar!/:3.8.3]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_144]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_144]
	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_144]
```

--- a/hazelcast/src/main/java/com/hazelcast/core/MapLoader.java
+++ b/hazelcast/src/main/java/com/hazelcast/core/MapLoader.java
@@ -40,8 +40,8 @@ public interface MapLoader<K, V> {
      * to obtain the value. Implementation can use any means of loading the given key;
      * such as an O/R mapping tool, simple SQL or reading a file etc.
      *
-     * @param key
-     * @return value of the key
+     * @param key, cannot be null
+     * @return value of the key, value cannot be null
      */
     V load(K key);
 
@@ -51,6 +51,9 @@ public interface MapLoader<K, V> {
      *
      * For any key in the input keys, there should be a single mapping in the resulting map. Also the resulting
      * map should not have any keys that are not part of the input keys.
+     *
+     * The given collection should not contain any <code>null</code> keys.
+     * The returned Map should not contain any <code>null</code> keys or values.
      *
      * @param keys keys of the values entries to load
      * @return map of loaded key-value pairs.
@@ -63,7 +66,9 @@ public interface MapLoader<K, V> {
      * {@link Closeable} interface in which case it will be closed once iteration is over.
      * This is intended for releasing resources such as closing a JDBC result set.
      *
-     * @return all the keys
+     * The returned Iterable should not contain any <code>null</code> keys.
+     *
+     * @return all the keys. Keys inside the Iterable cannot be null.
      */
     Iterable<K> loadAllKeys();
 }

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/MapKeyLoaderUtil.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/MapKeyLoaderUtil.java
@@ -32,6 +32,7 @@ import java.util.Map.Entry;
 import java.util.NoSuchElementException;
 
 import static com.hazelcast.config.MaxSizeConfig.MaxSizePolicy.PER_NODE;
+import static com.hazelcast.util.Preconditions.checkNotNull;
 
 public final class MapKeyLoaderUtil {
 
@@ -120,6 +121,8 @@ public final class MapKeyLoaderUtil {
 
         @Override
         public Entry<Integer, Data> apply(Data input) {
+            // Null-pointer here, in case of null key loaded by MapLoader
+            checkNotNull(input, "Key loaded by a MapLoader cannot be null.");
             Integer partition = partitionService.getPartitionId(input);
             return new MapEntrySimple<Integer, Data>(partition, input);
         }

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/operation/PutFromLoadAllOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/operation/PutFromLoadAllOperation.java
@@ -37,6 +37,7 @@ import java.util.List;
 
 import static com.hazelcast.util.CollectionUtil.isEmpty;
 import static com.hazelcast.util.Preconditions.checkFalse;
+import static com.hazelcast.util.Preconditions.checkNotNull;
 
 /**
  * Puts records to map which are loaded from map store by {@link com.hazelcast.core.IMap#loadAll}
@@ -65,13 +66,20 @@ public class PutFromLoadAllOperation extends MapOperation implements PartitionAw
         for (int i = 0; i < keyValueSequence.size(); i += 2) {
             Data key = keyValueSequence.get(i);
             Data dataValue = keyValueSequence.get(i + 1);
+
+            checkNotNull(key, "Key loaded by a MapLoader cannot be null.");
+
             // here object conversion is for interceptors.
             Object value = hasInterceptor ? mapServiceContext.toObject(dataValue) : dataValue;
             Object previousValue = recordStore.putFromLoad(key, value);
 
-            callAfterPutInterceptors(value);
+            // do not run interceptors in case the put was skipped due to null value
+            if (value != null) {
+                callAfterPutInterceptors(value);
+            }
             Record record = recordStore.getRecord(key);
             if (isPostProcessing(recordStore)) {
+                checkNotNull(record, "Value loaded by a MapLoader cannot be null.");
                 value = record.getValue();
             }
             publishEntryEvent(key, previousValue, value);

--- a/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/MapLoaderTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/MapLoaderTest.java
@@ -29,6 +29,7 @@ import com.hazelcast.core.MapStoreFactory;
 import com.hazelcast.instance.Node;
 import com.hazelcast.internal.partition.InternalPartition;
 import com.hazelcast.internal.partition.InternalPartitionService;
+import com.hazelcast.map.MapInterceptor;
 import com.hazelcast.map.impl.mapstore.writebehind.TestMapUsingMapStoreBuilder;
 import com.hazelcast.nio.Address;
 import com.hazelcast.query.SqlPredicate;
@@ -38,11 +39,15 @@ import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.TestHazelcastInstanceFactory;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
+import org.hamcrest.CoreMatchers;
+import org.junit.Rule;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
+import org.junit.rules.ExpectedException;
 import org.junit.runner.RunWith;
 
 import java.io.Serializable;
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
@@ -68,6 +73,9 @@ import static org.junit.Assert.assertNotNull;
 @RunWith(HazelcastSerialClassRunner.class)
 @Category({QuickTest.class, ParallelTest.class})
 public class MapLoaderTest extends HazelcastTestSupport {
+
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
 
     @Test
     public void testSenderAndBackupTerminates_AfterInitialLoad() throws InterruptedException {
@@ -132,6 +140,333 @@ public class MapLoaderTest extends HazelcastTestSupport {
         //assert loadAll with load all entries provided by the mapLoader
         map.loadAll(true);
         assertEquals(keysInMapStore, map.size());
+    }
+
+    @Test
+    public void testNullChecks_withMapStore_nullInKeys() {
+        String name = "testNullChecks_withMapStore";
+
+        int keysInMapStore = 10000;
+        Config config = new Config();
+        MapConfig mapConfig = config.getMapConfig(name);
+        MapStoreConfig mapStoreConfig = new MapStoreConfig();
+        mapStoreConfig.setEnabled(true);
+        mapStoreConfig.setImplementation(new MapLoaderTest.DummyMapLoader(keysInMapStore));
+        mapStoreConfig.setInitialLoadMode(MapStoreConfig.InitialLoadMode.EAGER);
+        mapConfig.setMapStoreConfig(mapStoreConfig);
+        HazelcastInstance instance = createHazelcastInstance(config);
+        IMap<String, String> map = instance.getMap(name);
+
+        final Set<String> keys = new HashSet<String>();
+        keys.add("key");
+        keys.add(null);
+
+        expectedException.expect(NullPointerException.class);
+        map.loadAll(keys, true);
+    }
+
+    @Test
+    public void testNullKey_loadAll() throws InterruptedException {
+        String name = "testNullIn_loadAllKeys";
+
+        Config config = new Config();
+
+        MapConfig mapConfig = config.getMapConfig(name);
+        MapStoreConfig mapStoreConfig = new MapStoreConfig();
+        mapStoreConfig.setEnabled(true);
+        mapStoreConfig.setImplementation(new MapLoader() {
+            @Override
+            public Object load(Object key) {
+                if (key.equals("1")) {
+                    return "1";
+                }
+                if (key.equals("2")) {
+                    return "2";
+                }
+                if (key.equals("3")) {
+                    return "3";
+                }
+                return null;
+            }
+
+            @Override
+            public Map loadAll(Collection keys) {
+                Map val = new HashMap();
+                if (keys.contains("1")) {
+                    val.put("1", "1");
+                }
+                if (keys.contains("2")) {
+                    val.put(null, "2");
+                }
+                if (keys.contains("3")) {
+                    val.put("3", "3");
+                }
+                return val;
+            }
+
+            @Override
+            public Iterable loadAllKeys() {
+                List keys = new ArrayList();
+                keys.add("1");
+                keys.add("2");
+                keys.add("3");
+                return keys;
+            }
+        });
+        mapStoreConfig.setInitialLoadMode(MapStoreConfig.InitialLoadMode.LAZY);
+        mapConfig.setMapStoreConfig(mapStoreConfig);
+        HazelcastInstance instance = createHazelcastInstance(config);
+        IMap<String, String> map = instance.getMap(name);
+
+        expectedException.expect(NullPointerException.class);
+        expectedException.expectMessage(CoreMatchers.is("Key loaded by a MapLoader cannot be null."));
+        map.size();
+
+        assertEquals(2, map.size());
+        assertEquals("1", map.get("1"));
+        assertEquals("2", map.get("2"));
+        assertEquals("3", map.get("3"));
+    }
+
+    @Test
+    public void testNullValue_loadAll() throws InterruptedException {
+        String name = "testNullIn_loadAllKeys";
+
+        Config config = new Config();
+
+        MapConfig mapConfig = config.getMapConfig(name);
+        MapStoreConfig mapStoreConfig = new MapStoreConfig();
+        mapStoreConfig.setEnabled(true);
+        mapStoreConfig.setImplementation(new MapLoader() {
+            @Override
+            public Object load(Object key) {
+                if (key.equals("1")) {
+                    return "1";
+                }
+                if (key.equals("2")) {
+                    return null;
+                }
+                if (key.equals("3")) {
+                    return "3";
+                }
+                return null;
+            }
+
+            @Override
+            public Map loadAll(Collection keys) {
+                Map val = new HashMap();
+                if (keys.contains("1")) {
+                    val.put("1", "1");
+                }
+                if (keys.contains("2")) {
+                    val.put("2", null);
+                }
+                if (keys.contains("3")) {
+                    val.put("3", "3");
+                }
+                return val;
+            }
+
+            @Override
+            public Iterable loadAllKeys() {
+                List keys = new ArrayList();
+                keys.add("1");
+                keys.add("2");
+                keys.add("3");
+                return keys;
+            }
+        });
+        mapStoreConfig.setInitialLoadMode(MapStoreConfig.InitialLoadMode.LAZY);
+        mapConfig.setMapStoreConfig(mapStoreConfig);
+        HazelcastInstance instance = createHazelcastInstance(config);
+        IMap<String, String> map = instance.getMap(name);
+
+        // THIS DOES NOT THROW ANY EXCEPTION
+        map.size();
+
+        assertEquals(2, map.size());
+        assertEquals("1", map.get("1"));
+        assertEquals(null, map.get("2"));
+        assertEquals("3", map.get("3"));
+    }
+
+    @Test
+    public void testNullValue_loadAll_withInterceptor() {
+        String name = "testNullIn_loadAllKeys";
+
+        Config config = new Config();
+
+        MapConfig mapConfig = config.getMapConfig(name);
+        MapStoreConfig mapStoreConfig = new MapStoreConfig();
+        mapStoreConfig.setEnabled(true);
+        mapStoreConfig.setImplementation(new MapLoader() {
+            @Override
+            public Object load(Object key) {
+                if (key.equals("1")) {
+                    return "1";
+                }
+                if (key.equals("2")) {
+                    return null;
+                }
+                if (key.equals("3")) {
+                    return "3";
+                }
+                return null;
+            }
+
+            @Override
+            public Map loadAll(Collection keys) {
+                Map val = new HashMap();
+                if (keys.contains("1")) {
+                    val.put("1", "1");
+                }
+                if (keys.contains("2")) {
+                    val.put("2", null);
+                }
+                if (keys.contains("3")) {
+                    val.put("3", "3");
+                }
+                return val;
+            }
+
+            @Override
+            public Iterable loadAllKeys() {
+                List keys = new ArrayList();
+                keys.add("1");
+                keys.add("2");
+                keys.add("3");
+                return keys;
+            }
+        });
+        mapStoreConfig.setInitialLoadMode(MapStoreConfig.InitialLoadMode.LAZY);
+        mapConfig.setMapStoreConfig(mapStoreConfig);
+        HazelcastInstance instance = createHazelcastInstance(config);
+        IMap<String, String> map = instance.getMap(name);
+
+        map.addInterceptor(new MyIntercepter());
+
+        expectedException.expect(NullPointerException.class);
+        expectedException.expectMessage(CoreMatchers.is("Value loaded by a MapLoader cannot be null."));
+        map.size();
+
+        assertEquals(2, map.size());
+        assertEquals("1", map.get("1"));
+        assertEquals(null, map.get("2"));
+        assertEquals("3", map.get("3"));
+    }
+
+    private static class MyIntercepter implements MapInterceptor, Serializable {
+        @Override
+        public Object interceptGet(Object value) {
+            return null;
+        }
+
+        @Override
+        public void afterGet(Object value) {
+
+        }
+
+        @Override
+        public Object interceptPut(Object oldValue, Object newValue) {
+            return null;
+        }
+
+        @Override
+        public void afterPut(Object value) {
+
+        }
+
+        @Override
+        public Object interceptRemove(Object removedValue) {
+            return null;
+        }
+
+        @Override
+        public void afterRemove(Object value) {
+
+        }
+    }
+
+    @Test
+    public void testNullKey_loadAllKeys() throws InterruptedException {
+        String name = "testNullIn_loadAllKeys";
+
+        Config config = new Config();
+
+        MapConfig mapConfig = config.getMapConfig(name);
+        MapStoreConfig mapStoreConfig = new MapStoreConfig();
+        mapStoreConfig.setEnabled(true);
+        mapStoreConfig.setImplementation(new MapLoader() {
+            @Override
+            public Object load(Object key) {
+                if (key.equals("1")) {
+                    return "1";
+                }
+                if (key.equals("2")) {
+                    return "2";
+                }
+                if (key.equals("3")) {
+                    return "3";
+                }
+                return null;
+            }
+
+            @Override
+            public Map loadAll(Collection keys) {
+                Map val = new HashMap();
+                if (keys.contains("1")) {
+                    val.put("1", "1");
+                }
+                if (keys.contains("2")) {
+                    val.put("2", "2");
+                }
+                if (keys.contains("3")) {
+                    val.put("3", "3");
+                }
+                return val;
+            }
+
+            @Override
+            public Iterable loadAllKeys() {
+                List keys = new ArrayList();
+                keys.add("1");
+                keys.add(null);
+                keys.add("3");
+                return keys;
+            }
+        });
+        mapStoreConfig.setInitialLoadMode(MapStoreConfig.InitialLoadMode.LAZY);
+        mapConfig.setMapStoreConfig(mapStoreConfig);
+        HazelcastInstance instance = createHazelcastInstance(config);
+        IMap<String, String> map = instance.getMap(name);
+
+        expectedException.expect(NullPointerException.class);
+        expectedException.expectMessage(CoreMatchers.is("Key loaded by a MapLoader cannot be null."));
+        map.size();
+
+        assertEquals(2, map.size());
+        assertEquals("1", map.get("1"));
+        assertEquals(null, map.get("2"));
+        assertEquals("3", map.get("3"));
+    }
+
+    @Test
+    public void testNullChecks_withMapStore_nullKeys() {
+        String name = "testNullChecks_withMapStore";
+        int keysInMapStore = 10000;
+
+        Config config = new Config();
+        MapConfig mapConfig = config.getMapConfig(name);
+        MapStoreConfig mapStoreConfig = new MapStoreConfig();
+        mapStoreConfig.setEnabled(true);
+        mapStoreConfig.setImplementation(new MapLoaderTest.DummyMapLoader(keysInMapStore));
+        mapStoreConfig.setInitialLoadMode(MapStoreConfig.InitialLoadMode.EAGER);
+        mapConfig.setMapStoreConfig(mapStoreConfig);
+        HazelcastInstance instance = createHazelcastInstance(config);
+        IMap<String, String> map = instance.getMap(name);
+
+        expectedException.expect(NullPointerException.class);
+        map.loadAll(null, true);
     }
 
     private HazelcastInstance getInstanceForAddress(HazelcastInstance[] instances, Address address) {

